### PR TITLE
nixos/update-users-groups: let hashedPassword take precedence over initialHashedPassword

### LINKED ISSUE #208766: Ephemeral rootfs setup:  root user doesn't get password set
### Describe the bug
On my server I followed @grahamc 's guide on setting up an ephemeral rootfs server (see his blog post [https://grahamc.com/blog/erase-your-darlings](https://grahamc.com/blog/erase-your-darlings) ). However I'm having problem giving the root user a password. Whenever I look at the shadow file, it does not contain any password hash.

So I wanted to test this out now in a VM and I have the same results. For easier testing I did write a small bash script to automate the partitioning and zfs setup as well as  a basic configuration file to be used with it.

### Steps To Reproduce
1. Setup a VM (100GB virtual disk), copy `installer` script and the `configuration.nix` (see blow) to the same folder
2. Run the `installer` script. This will just do the partitioning and setup the zpool and copy the configuration.nix to its proper location
3. Run `nixos-install` and then reboot

Password for both (user `root` and user `hyper`) is: `thisisatest`

### Expected behavior
After reboot you should be able to login as `root` user with the password above. Also you should see the hashed password in the `/etc/shadow` file.

However you can't login and the hashed password entry for the root user in the shadow file is empty. For the user `hyper` login works.

The `/etc/shadow` looks after reboot like this:
```
[...]
root:!:1::::::
[...]
hyper:$6$t/ER3ncagRvDGY/d$IL9OEPymL8.t97V/PKOA4CRX9wWsvqcHp.6oGAnfRACB54FqlSTa7FLlQqKrcqr2uH.iohlJnNgHnZdm2U2A5[...]
```

### Additional context
I tried with omitting the `mutableUserscd` option and setting it to false. Neither worked.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.83, NixOS, 23.05 (Stoat), 23.05pre435935.04f574a1c0f`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.12.0`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### Files
`Installer`:
```
#!/usr/bin/env bash

# Restore partition layout
echo 'label: gpt
label-id: 6B6AEDAF-736B-A145-8753-D9C27A82D7F2
device: /dev/sda
unit: sectors
first-lba: 34
last-lba: 209715166
sector-size: 512

/dev/sda1 : start=        2048, size=       20480, type=21686148-6449-6E6F-744E-656564454649, uuid=B88A0C6F-1835-4D9B-9AF3-C71ACDC7F9BF, name="BIOS boot partition"
/dev/sda2 : start=       22528, size=     2097152, type=C12A7328-F81F-11D2-BA4B-00A0C93EC93B, uuid=6C90F2B6-9D19-43FC-8C99-C6D1717A2626, name="EFI system partition"
/dev/sda3 : start=     2119680, size=   207593472, type=0FC63DAF-8483-4772-8E79-3D69D8477DE4, uuid=CF83F42F-BE1A-4366-8CB3-8E7383A50491, name="Linux filesystem"' > part_table

sfdisk --force /dev/sda < part_table

# Format EFI partition
mkfs.fat -F 32 /dev/sda2

# Create Pool
zpool destroy tankTest
zpool create -f -o ashift=12 -o altroot="/mnt" -O mountpoint=none -O compression=zstd-2 -O atime=off -O xattr=sa -O acltype=posixacl -O normalization=formD -O relatime=on "tankTest" "/dev/sda3"

# Create Base Datasets
zfs create tankTest/encZFS
zfs create -o mountpoint=legacy tankTest/encZFS/Nixos

# Create Perma Datasets
zfs create tankTest/encZFS/Nixos/p
zfs create tankTest/encZFS/Nixos/p/data
zfs create tankTest/encZFS/Nixos/p/etc
zfs create tankTest/encZFS/Nixos/p/etc/nixos
zfs create tankTest/encZFS/Nixos/p/home
zfs create tankTest/encZFS/Nixos/p/root

# Create Volatile Datasets
zfs create tankTest/encZFS/Nixos/v
zfs create tankTest/encZFS/Nixos/v/rootfs
zfs create tankTest/encZFS/Nixos/v/nix

# Create Blank Snapshot
zfs snapshot tankTest/encZFS/Nixos/v/rootfs@blank

# Mount Partitions and Datasets
mount -t zfs tankTest/encZFS/Nixos/v/rootfs /mnt
mkdir /mnt/boot
mount /dev/sda2 /mnt/boot
mkdir /mnt/data
mount -t zfs tankTest/encZFS/Nixos/p/data /mnt/data
mkdir -p /mnt/etc/nixos
mount -t zfs tankTest/encZFS/Nixos/p/etc/nixos /mnt/etc/nixos
mkdir /mnt/home
mount -t zfs tankTest/encZFS/Nixos/p/home /mnt/home
mkdir /mnt/root
mount -t zfs tankTest/encZFS/Nixos/p/root /mnt/root
mkdir /mnt/nix
mount -t zfs tankTest/encZFS/Nixos/v/nix /mnt/nix

# Create Nixos Configurations
nixos-generate-config --root /mnt

# Copy configuration.nix
cp configuration.nix /mnt/etc/nixos/configuration.nix
```

`configuration.nix`:
```
# Edit this configuration file to define what should be installed on
# your system.  Help is available in the configuration.nix(5) man page
# and in the NixOS manual (accessible by running ‘nixos-help’).

{ config, pkgs, lib, ... }:

{
    imports =
        [ # Include the results of the hardware scan.
            ./hardware-configuration.nix
        ];

    # Use the systemd-boot EFI boot loader.
    boot.loader.systemd-boot.enable = true;
    boot.loader.efi.canTouchEfiVariables = true;

    # Rollback RootFS Dataset
    boot.initrd.postDeviceCommands = lib.mkAfter ''
        zfs rollback -r tankTest/encZFS/Nixos/v/rootfs@blank
    '';

    # Add more filesystems
    boot.supportedFilesystems = [ "zfs" ];
    boot.zfs.enableUnstable = true;
    boot.zfs.devNodes = "/dev";
    services.zfs.autoScrub = {
        enable = true;
        interval = "monthly";
        pools = [ ]; # List of ZFS pools to periodically scrub. If empty, all pools will be scrubbed.
    };

    # Trust hydra. Needed for one-click installations.
    nix.settings.trusted-substituters = [ "http://hydra.nixos.org" ];

    networking = {
        useDHCP = true;
        nameservers = [ "10.10.10.1" "8.8.8.8" "8.8.4.4" ];
        hostName = "testi"; # Define your hostname.
        hostId = "100000cc";
        firewall = {
            enable = false;
            allowPing = true;
	};
    };



    # Enable dbus
    services.dbus.enable = true;

    # Select internationalisation properties.
    console.font = "Lat2-Terminus16";
    console.keyMap ="sg-latin1";
    i18n = {
        defaultLocale = "en_US.UTF-8";
        supportedLocales = [ "all" ];
    };

    # Enable the OpenSSH daemon.
    services.openssh = {
        enable = true;
        hostKeys = [
            {   path = "/data/ssh_host_ed25519_key";
                type = "ed25519"; }
            {   path = "/data/ssh_host_rsa_key";
                type = "rsa";
                bits = 4096; }
        ];
        permitRootLogin = "yes";
        ports = [ 22 23 ];
    };

    # Enable nscd
    services.nscd = {
        enable = true;
    };

    # Enable ntp or rather timesyncd
    services.timesyncd = {
        enable = true;
        servers = [ "0.ch.pool.ntp.org" "1.ch.pool.ntp.org" "2.ch.pool.ntp.org" "3.ch.pool.ntp.org" ];
    };

    # Time.
    time.timeZone = "Europe/Zurich";

    # Enable sudo
    security.sudo = {
        enable = true;
        wheelNeedsPassword = true;
    };

    # Define a user account. Don't forget to set a password with ‘passwd’.
    users.defaultUserShell = "/run/current-system/sw/bin/bash";
    users.mutableUsers = false;
    users.users = {
        root = {
            hashedPassword = "$6$t/ER3ncagRvDGY/d$IL9OEPymL8.t97V/PKOA4CRX9wWsvqcHp.6oGAnfRACB54FqlSTa7FLlQqKrcqr2uH.iohlJnNgHnZdm2U2A5/";
        };
        hyper = {
            isNormalUser = true;    # creates home, adds to group users, sets default shell
            createHome = false;
            home = "/home/hyper";
            description = "hyper";
            extraGroups = [ "networkmanager" "vboxusers" "wheel" "audio" "cdrom" "kvm" ]; # wheel is for the sudo group
            uid = 1000;
            hashedPassword = "$6$t/ER3ncagRvDGY/d$IL9OEPymL8.t97V/PKOA4CRX9wWsvqcHp.6oGAnfRACB54FqlSTa7FLlQqKrcqr2uH.iohlJnNgHnZdm2U2A5/";
        };
    };

    # Enable Locate
    services.locate.enable = true;

    # Bash completion
    programs.bash.enableCompletion = true;

    # Setup nano
    programs.nano.nanorc = ''
        set nowrap
        set tabstospaces
        set tabsize 4
        set constantshow
        # include /usr/share/nano/sh.nanorc
    '';



    system.stateVersion = "23.05"; # Did you read the comment?

    nixpkgs.config.allowUnfree = true;

    environment.systemPackages = with pkgs; [
        bash-completion
    	curl
    	htop
        jq
    	mc
    	tmux
    ];

}
```

### PR DESCRIPTION
**Pull request by lilyinstarlight**
###### Description of changes

Without this change, users that have both `initialHashedPassword` and `hashedPassword` set will have `initialHashedPassword` take precedence, but only for the first time `/etc/passwd` is generated. After that, `hashedPassword` takes precedence. This is surprising behavior as it would generally be expected for `hashedPassword` to win if both are set.

This wouldn't be a noticeable problem (and an assert could just be made instead) if the users-groups module did not also default the `root.intialHashedPassword` value to `!`, to prevent login by default. That means that users who set `root.hashedPassword` and use an ephemeral rootfs (i.e. `/etc/passwd` is created every boot) are not able to log in to the root account by default, unless they switch to a new generation during the same boot (i.e. `/etc/passwd` already exists and `hashedPassword` is then used instead of `initialHashedPassword`) or they set `root.initialHashedPassword = null` (which is unintuitive and seems redundant).

Closes #208766

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lilyinstarlight** [comment]
@ofborg build nixosTests.shadow

**Review by dasJ — APPROVED** [review]
The change seems reasonable from a UX standpoint and the code looks good as well.

The only nit I have (but I am aware how annoying these kinds of requests are) is that it would be nice if the NixOS options would somehow document this. Something like `Takes precedence over X` and/or `Overridden by X` would make it intuitive **and** documented in case people wonder ;)

**Comment by lilyinstarlight** [comment]
> The only nit I have (but I am aware how annoying these kinds of requests are) is that it would be nice if the NixOS options would somehow document this. Something like `Takes precedence over X` and/or `Overridden by X` would make it intuitive **and** documented in case people wonder ;)

That's fair! I did consider this as a good idea, but I was feeling a tad lazy at the time I wrote this PR

I'll see if I can adjust the option descriptions later today to clarify precedence

**Comment by lilyinstarlight** [comment]
I also updated the `initialPassword` documentation too while I was at it, since it looks like `password` does take precedence over `initialPassword` already but doesn't give a clear precedence order in the description

**Review by dasJ — APPROVED** [review]

File: nixos/modules/config/users-groups.nix [hunk 1]
```diff
@@ -273,6 +273,9 @@ let
           {command}`passwd` command. Otherwise, it's
           equivalent to setting the {option}`hashedPassword` option.
 
+          Note that the {option}`hashedPassword` option will still
+          override this option if both are set.
```
Comment (winterqt) [inline nixos/modules/config/users-groups.nix:277]: I'm unsure where the "still" comes into play here, which reads as if it's mentioned before this line.

```suggestion
          Note that the {option}`hashedPassword` option will
          override this option if both are set.
```

File: nixos/modules/config/users-groups.nix [hunk 2]
```diff
@@ -291,6 +294,9 @@ let
           is world-readable in the Nix store, so it should only be
           used for guest accounts or passwords that will be changed
           promptly.
+
+          Note that the {option}`password` option will still override
+          this option if both are set.
```
Comment (winterqt) [inline nixos/modules/config/users-groups.nix:299]: Ditto.

```suggestion
          Note that the {option}`password` option will override
          this option if both are set.
```

**Review by winterqt — CHANGES_REQUESTED** [review]
Just a small wording change, good to go otherwise.

@dasJ Think this is worth mentioning in the release notes?

File: nixos/modules/config/users-groups.nix (hunk 1 above)
Comment (lilyinstarlight) [reply nixos/modules/config/users-groups.nix]: Fixed, thanks!

**Review by winterqt — APPROVED** [review]
🐈

**Pull request merged**